### ISSUE TITLE
Flow checker doesn't complain

### ISSUE THREAD

**Issue by NN---** [Microsoft/TypeScript#13810]
Calling class function can change the variable and making possible to call function incorrectly.

**TypeScript Version:**  2.1.5
strictNullChecks: true

**Code**

```ts
class A {
	x: number | undefined;

	f() {
		this.x = 1;
		this.h();
		this.g(this.x); // Pass 'undefined'
	}

	g(q: number) { }

	h() { this.x = undefined; }
}
```

**Expected behavior:**
Complain that this.g can get undefined.

**Actual behavior:**
Compiles fine.

**Comment by HerringtonDarkholme** [comment]
https://github.com/Microsoft/TypeScript/issues/9998

**Comment by devdoomari** [comment]
I don't know much about TS control flow analysis, but

```
class A {
	x: number | undefined;

	f() {
		this.x = 1;  <----- does this affect control-flow check?
		this.h();
		this.g(this.x); // Pass 'undefined'
	}

	g(q: number) { }

	h() { this.x = undefined; }
}
```

does this.x = 1 affect TS control flow analysis?
